### PR TITLE
AMD Optimized CPU Libraries: add v4.0

### PR DESCRIPTION
**Pull request by amd-toolchain-support**
AOCL is a set of numerical libraries optimized for AMD processors based on the AMD "Zen" core architecture and generations.  Supported processor families are AMD EPYC™, AMD Ryzen™, and AMD Ryzen™ Threadripper™ processors. The tuned implementations of industry-standard math libraries enable rapid development of scientific and high performance computing applications.

What's in AOCL 4.0:
1. amdblis LPGEMM variants with post-ops support AMD "Zen4" support for BLIS

2. amdlibflame Upgrade to LAPACK 3.10.1 specification Improvements in a few more variants of SVD and Eigen Value routines Multithread support enabled for selected APIs

3. amdfftw AVX-512 enablement of DFT kernels AVX-512 optimization of copy and transpose routines

5. amdlibm Black & Scholes support (logf, expf, erff, both scalar and vector) AVX-512 variants of vector functions

6. aocl-sparse New Iterative Solver APIs AVX-512 support for SPMV API

7. amdscalapack Upgrade to Netlib ScaLAPACK 2.2.0

Change-Id: Ib850f3ec0bccc9cf1c35747cd56302951dc88696

### CODE REVIEW

File: var/spack/repos/builtin/packages/amdlibm/package.py [hunk 1]
```diff
@@ -41,29 +45,26 @@ class Amdlibm(SConsPackage):
...
-        when="@2.2:3.0",
-        msg="amdlibm 2.2 and 3.0 versions are not supported with AOCC 3.2.0",
-    )
+    conflicts("%gcc@:9.1.0", msg="Minimum supported GCC version is 9.2.0")
+    conflicts("%gcc@12.2.0:", msg="Maximum supported GCC version is 12.1.0")
+    conflicts("%clang@9:", msg="Minimum supported Clang version is 9.0.0")
+    conflicts("%aocc@3.2.0", msg="dependency on python@3.6.2")
 
-    def build_args(self, spec, prefix):
+    def build_args(self, *args, **kwargs):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/amdlibm/package.py]: Why did you introduce this change? Can you please revert and use a more static signature signature?

File: var/spack/repos/builtin/packages/aocl-sparse/package.py [hunk 2]
```diff
@@ -77,6 +90,13 @@ def cmake_args(self):
             ]
         )
 
+        args.extend(
+            [
+                self.define_from_variant("BUILD_WITH_AVX512", "AVX"),
+                "-DUSE_AVX512:BOOL=%s" % ("ON" if self.run_tests else "OFF"),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/aocl-sparse/package.py]:
```suggestion
                self.define("USE_AVX512", self.run_tests),
```

**Review by alalazo — CHANGES_REQUESTED** [review]
Just a couple of minor change requests

File: var/spack/repos/builtin/packages/amdlibm/package.py (hunk 1 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/amdlibm/package.py]: amdlibm builds are failing at the install stage with below error:

--snip--
~/spack/etc/spack/licenses/amdlibm/AMDLIBM_EULA.pdf" 25L, 842B written
    ==> amdlibm: Executing phase: 'build'
    ==> amdlibm: Executing phase: 'install'
    ==> Error: TypeError: Amdlibm.install_args() missing 2 required positional arguments: 'spec' and 'prefix'
    
    /home/amd-toolchain-support/spack/lib/spack/spack/builder.py:139, in __forward:
      >>    139        def __forward(self, *args, **kwargs):
            140            return self.__getattr__(fn_name)(*args, **kwargs)
--snip--

Comment (alalazo) [reply var/spack/repos/builtin/packages/amdlibm/package.py]: Can you revert? The error seems to point to another function. Also, it's bad practice to have catch all signatures to work around errors when passing parameters. If there's still an error it means there's a bug somewhere in Spack we need to fix.

Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/amdlibm/package.py]: Now the build issue is not seen, so reverted the change.

--snip--
-    def build_args(self, *args, **kwargs):
+    def build_args(self, spec, prefix):
--snip--

**Review by alalazo — DISMISSED** [review]

File: var/spack/repos/builtin/packages/aocl-sparse/package.py [hunk 3]
```diff
@@ -35,7 +42,17 @@ class AoclSparse(CMakePackage):
...
     variant("ilp64", default=False, description="Build with ILP64 support")
+    variant("examples", default=False, description="Build sparse examples")
+    variant("unit_tests", default=False, description="Build sparse unit tests")
+    variant("benchmarks", default=False, description="Build Build benchmarks")
+    variant(
+        "avx",
+        default=False,
+        when="target=zen4:",
+        description="Build with AVX512 support",
+    )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/aocl-sparse/package.py]: Can't we avoid the variant and just infer from the target?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/aocl-sparse/package.py]: Or is this feature truly optional for those targets?

**Review by tldahlgren — DISMISSED** [review]
Confirmed all of the new sha256.  Please clarify the `avx` variant is actually optional on those targets.

File: var/spack/repos/builtin/packages/aocl-sparse/package.py (hunk 3 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/aocl-sparse/package.py]: Yes, this feature is truly optional as it is currently experimental.

Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/aocl-sparse/package.py]: Below is an updated snippet

--snip--
     ```
variant(
         "avx",
         default=False,
-        when="target=zen4:",
-        description="Build with AVX512 support",
+        when="@4.0: target=zen4:",
+        description="Enable experimental AVX512 support",
     )
```
--snip--

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]

**Pull request merged**